File: src/aria.go
package src

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

//go:embed aria.js
var ariaJS string

const lookupJS = `(id) => {
	const ref = window.__bmcp && window.__bmcp.els.get(id);
	const el = ref && ref.deref();
	return el && el.isConnected ? el : null;
}`

type ariaNode struct {
	Role        string      `json:"r"`
	Name        string      `json:"n"`
	ID          string      `json:"id"`
	Value       string      `json:"v"`
	Placeholder string      `json:"ph"`
	States      []string    `json:"st"`
	Level       int         `json:"lvl"`
	Href        string      `json:"href"`
	TestID      string      `json:"tid"`
	Text        string      `json:"t"`
	Children    []*ariaNode `json:"c"`
}

type ariaSnapshot struct {
	Tree    []*ariaNode `json:"tree"`
	Next    int         `json:"next"`
	URL     string      `json:"url"`
	Title   string      `json:"title"`
	ScrollX int         `json:"sx"`
	ScrollY int         `json:"sy"`
	ScrollW int         `json:"sw"`
	ScrollH int         `json:"sh"`
}

func parseAriaSnapshot(raw any) (*ariaSnapshot, error) {
	str, ok := raw.(string)
	if !ok {
		return nil, fmt.Errorf("unexpected aria dump result %T", raw)
	}
	var snap ariaSnapshot
	if err := json.Unmarshal([]byte(str), &snap); err != nil {
		return nil, err
	}
	snap.Tree = compactNodes(snap.Tree)
	return &snap, nil
}

// compactNodes removes wrappers that add lines without adding information.
func compactNodes(nodes []*ariaNode) []*ariaNode {
	out := make([]*ariaNode, 0, len(nodes))
	for i, n := range nodes {
		n.Children = compactNodes(n.Children)
		if n.isEmpty() {
			continue
		}
		// Label text right next to the control it names is already in the control's name.
		if n.Role == "" && ((i > 0 && nodes[i-1].Name == n.Text) || (i+1 < len(nodes) && nodes[i+1].Name == n.Text)) {
			continue
		}
		if n.Name != "" {
			// Text that is already part of the name adds nothing.
			kept := n.Children[:0]
			for _, c := range n.Children {
				if c.Role != "" || !strings.Contains(n.Name, c.Text) {
					kept = append(kept, c)
				}
			}
			n.Children = kept
		}
		wrapper := n.Role == "listitem" || n.Role == "cell"
		if wrapper && n.ID == "" && n.Name == "" && len(n.States) == 0 && n.TestID == "" && len(n.Children) == 1 {
			out = append(out, n.Children[0])
			continue
		}
		out = append(out, n)
	}
	return out
}

func (n *ariaNode) isEmpty() bool {
	if n.Role == "" {
		return n.Text == ""
	}
	return n.ID == "" && n.Name == "" && n.Value == "" && len(n.States) == 0 && n.TestID == "" &&
		len(n.Children) == 0 && n.Role != "separator" && n.Role != "iframe"
}

func (n *ariaNode) line() string {
	if n.Role == "" {
		return strconv.Quote(n.Text)
	}

	var b strings.Builder
	if n.ID != "" {
		b.WriteString("[" + n.ID + "] ")
	}
	b.WriteString(n.Role)
	if n.Level > 0 {
		fmt.Fprintf(&b, "[%d]", n.Level)
	}
	if n.Name != "" {
		b.WriteString(" " + strconv.Quote(n.Name))
	}
	if n.Value != "" {
		b.WriteString(" value=" + strconv.Quote(n.Value))
	}
	if n.Placeholder != "" {
		b.WriteString(" placeholder=" + strconv.Quote(n.Placeholder))
	}
	if n.Href != "" {
		b.WriteString(" -> " + n.Href)
	}
	if n.TestID != "" {
		b.WriteString(" testid=" + n.TestID)
	}
	for _, s := range n.States {
		b.WriteString(" " + s)
	}
	return b.String()
}

// tableRowLine renders a row of plain cells on a single line.
func (n *ariaNode) tableRowLine() (string, bool) {
	if n.Role != "row" || n.ID != "" || len(n.Children) == 0 {
		return "", false
	}
	cells := make([]string, 0, len(n.Children))
	for _, c := range n.Children {
		isCell := c.Role == "cell" || c.Role == "columnheader" || c.Role == "rowheader"
		if !isCell || c.ID != "" || len(c.Children) > 0 || len(c.States) > 0 {
			return "", false
		}
		cells = append(cells, c.Name)
	}
	return "row: " + strings.Join(cells, " | "), true
}

func renderNodes(b *strings.Builder, nodes []*ariaNode, depth int) {
	for _, n := range nodes {
		b.WriteString(strings.Repeat("  ", depth))
		if row, ok := n.tableRowLine(); ok {
			b.WriteString(row + "\n")
			continue
		}
		b.WriteString(n.line() + "\n")
		renderNodes(b, n.Children, depth+1)
	}
}

var idQueryRe = regexp.MustCompile(`^\[?e\d+\]?$`)

// filterNodes keeps nodes matching the query (with their full subtree) and
// the ancestors needed to place them in context.
func filterNodes(nodes []*ariaNode, query string) []*ariaNode {
	query = strings.ToLower(strings.TrimSpace(query))
	idQuery := ""
	if idQueryRe.MatchString(query) {
		idQuery = strings.Trim(query, "[]")
	}

	var walk func([]*ariaNode) []*ariaNode
	walk = func(nodes []*ariaNode) []*ariaNode {
		var out []*ariaNode
		for _, n := range nodes {
			matched := false
			if idQuery != "" {
				matched = n.ID == idQuery
			} else {
				matched = strings.Contains(strings.ToLower(n.line()), query)
			}
			if matched {
				out = append(out, n)
				continue
			}
			if kept := walk(n.Children); len(kept) > 0 {
				cp := *n
				cp.Children = kept
				out = append(out, &cp)
			}
		}
		return out
	}
	return walk(nodes)
}

func renderAria(snap *ariaSnapshot, query string, maxChars int) string {
	nodes := snap.Tree
	if query != "" {
		nodes = filterNodes(nodes, query)
		if len(nodes) == 0 {
			return noMatchMessage(snap.Tree, query)
		}
	}

	var b strings.Builder
	renderNodes(&b, nodes, 0)
	out := b.String()
	if out == "" {
		return "(page has no visible content)\n"
	}
	if maxChars > 0 && len(out) > maxChars {
		cut := strings.LastIndexByte(out[:maxChars], '\n') + 1
		out = out[:cut] + fmt.Sprintf("… truncated %d more characters, use an aria query to narrow the dump down\n", len(out)-cut)
	}
	return out
}

// noMatchMessage explains an empty query result and lists the roles on the
// page so the next query can target something that exists.
func noMatchMessage(nodes []*ariaNode, query string) string {
	q := strings.ToLower(strings.TrimSpace(query))
	if idQueryRe.MatchString(q) {
		return fmt.Sprintf("(no element with id %q on the page)\n", strings.Trim(q, "[]"))
	}

	var roles []string
	var walk func([]*ariaNode)
	walk = func(nodes []*ariaNode) {
		for _, n := range nodes {
			if n.Role != "" && !slices.Contains(roles, n.Role) {
				roles = append(roles, n.Role)
			}
			walk(n.Children)
		}
	}
	walk(nodes)
	slices.Sort(roles)

	msg := fmt.Sprintf("(nothing matched query %q, the query matches the text of dump lines (role, name, value, link target) case insensitive", query)
	if len(roles) > 0 {
		msg += ", roles on this page: " + strings.Join(roles, ", ")
	}
	return msg + ")\n"
}

File: src/aria.js
// Builds a compact accessibility tree of the current document.
// Called with the lowest element id the session may hand out, so ids never
// repeat across navigations within one session.
(startId) => {
	const S = window.__bmcp || (window.__bmcp = { next: startId, ids: new WeakMap(), els: new Map() });
	if (S.next < startId) S.next = startId;

	const MAX_NAME = 150;
	const MAX_TEXT = 400;
	const MAX_OPTIONS = 50;

	const SKIP = new Set(['script', 'style', 'noscript', 'template', 'head', 'meta', 'link', 'title', 'base']);
	const INTERACTIVE = new Set([
		'link', 'button', 'textbox', 'searchbox', 'checkbox', 'radio', 'combobox', 'listbox', 'slider',
		'spinbutton', 'switch', 'tab', 'menuitem', 'menuitemcheckbox', 'menuitemradio', 'treeitem', 'option', 'clickable',
	]);
	const NAME_FROM_CONTENT = new Set([
		'button', 'link', 'heading', 'tab', 'menuitem', 'menuitemcheckbox', 'menuitemradio', 'option', 'treeitem',
		'cell', 'columnheader', 'rowheader', 'checkbox', 'radio', 'switch', 'tooltip', 'clickable', 'legend', 'caption',
	]);
	const INPUT_ROLES = {
		button: 'button', submit: 'button', reset: 'button', image: 'button', checkbox: 'checkbox', radio: 'radio',
		range: 'slider', number: 'spinbutton', search: 'searchbox', color: 'button', file: 'button',
	};
	const SECTIONING = 'article, aside, main, nav, section';

	const collapse = (s) => (s || '').replace(/\s+/g, ' ').trim();
	const clip = (s, n) => (s.length > n ? s.slice(0, n - 1) + '…' : s);

	function implicitRole(el) {
		const tag = el.localName;
		switch (tag) {
			case 'a': case 'area': return el.hasAttribute('href') ? 'link' : null;
			case 'button': case 'summary': return 'button';
			case 'input': {
				const t = (el.getAttribute('type') || 'text').toLowerCase();
				if (t === 'hidden') return null;
				return INPUT_ROLES[t] || 'textbox';
			}
			case 'select': return el.multiple || el.size > 1 ? 'listbox' : 'combobox';
			case 'textarea': return 'textbox';
			case 'h1': case 'h2': case 'h3': case 'h4': case 'h5': case 'h6': return 'heading';
			case 'img': return el.getAttribute('alt') === '' ? null : 'img';
			case 'nav': return 'navigation';
			case 'main': return 'main';
			case 'header': return el.parentElement && el.parentElement.closest(SECTIONING) ? null : 'banner';
			case 'footer': return el.parentElement && el.parentElement.closest(SECTIONING) ? null : 'contentinfo';
			case 'aside': return 'complementary';
			case 'form': return 'form';
			case 'section': return el.hasAttribute('aria-label') || el.hasAttribute('aria-labelledby') ? 'region' : null;
			case 'article': return 'article';
			case 'ul': case 'ol': case 'menu': return 'list';
			case 'li': return 'listitem';
			case 'table': return 'table';
			case 'tr': return 'row';
			case 'td': return 'cell';
			case 'th': return el.getAttribute('scope') === 'row' ? 'rowheader' : 'columnheader';
			case 'caption': return 'caption';
			case 'dialog': return 'dialog';
			case 'fieldset': case 'details': return 'group';
			case 'legend': return 'legend';
			case 'option': return 'option';
			case 'progress': return 'progressbar';
			case 'meter': return 'meter';
			case 'hr': return 'separator';
			case 'iframe': return 'iframe';
		}
		return null;
	}

	function textById(ids) {
		return ids.split(/\s+/).map((id) => {
			const el = document.getElementById(id);
			return el ? collapse(el.innerText || el.textContent) : '';
		}).filter(Boolean).join(' ');
	}

	function nameOf(el, role) {
		const tag = el.localName;
		const lb = el.getAttribute('aria-labelledby');
		if (lb) {
			const n = textById(lb);
			if (n) return [n, false];
		}
		const al = collapse(el.getAttribute('aria-label'));
		if (al) return [al, false];
		if (tag === 'input' || tag === 'select' || tag === 'textarea') {
			const t = (el.getAttribute('type') || '').toLowerCase();
			if (['button', 'submit', 'reset'].includes(t)) return [collapse(el.value) || (t === 'submit' ? 'Submit' : t === 'reset' ? 'Reset' : ''), false];
			if (t === 'image') return [collapse(el.alt || el.value), false];
			if (el.labels && el.labels.length) {
				const n = Array.from(el.labels).map((l) => collapse(l.innerText || l.textContent)).filter(Boolean).join(' ');
				if (n) return [n, false];
			}
			return [collapse(el.getAttribute('title') || el.getAttribute('placeholder')), false];
		}
		if (tag === 'img' || tag === 'area') return [collapse(el.getAttribute('alt') || el.getAttribute('title')), false];
		if (tag === 'fieldset') {
			const lg = el.querySelector(':scope > legend');
			if (lg) return [collapse(lg.innerText), false];
		}
		if (tag === 'table') {
			const cp = el.querySelector(':scope > caption');
			if (cp) return [collapse(cp.innerText), false];
		}
		if (NAME_FROM_CONTENT.has(role)) {
			const n = collapse(el.innerText || el.textContent);
			if (n) return [n, true];
		}
		return [collapse(el.getAttribute('title')), false];
	}

	function idFor(el) {
		let id = S.ids.get(el);
		if (!id) {
			id = 'e' + S.next++;
			S.ids.set(el, id);
		}
		S.els.set(id, new WeakRef(el));
		return id;
	}

	function childNodesOf(el) {
		if (el.shadowRoot) return el.shadowRoot.childNodes;
		if (el.localName === 'slot') {
			const assigned = el.assignedNodes({ flatten: true });
			if (assigned.length) return assigned;
		}
		return el.childNodes;
	}

	const BR = { br: true };

	function walkChildren(el, parentStyle) {
		const out = [];
		for (const k of childNodesOf(el)) {
			for (const item of walk(k, parentStyle)) out.push(item);
		}
		return out;
	}

	function states(el, role, cs) {
		const st = [];
		const aria = (n) => el.getAttribute('aria-' + n);
		if (el.disabled || aria('disabled') === 'true') st.push('disabled');
		if (role === 'checkbox' || role === 'radio' || role === 'switch' || role === 'menuitemcheckbox' || role === 'menuitemradio') {
			if (el.indeterminate || aria('checked') === 'mixed') st.push('mixed');
			else if (el.checked === true || aria('checked') === 'true') st.push('checked');
		}
		const exp = aria('expanded');
		if (exp === 'true') st.push('expanded');
		else if (exp === 'false') st.push('collapsed');
		if (el.localName === 'details') st.push(el.open ? 'expanded' : 'collapsed');
		if (aria('selected') === 'true' || (el.localName === 'option' && el.selected)) st.push('selected');
		if (aria('pressed') === 'true') st.push('pressed');
		if (aria('current') && aria('current') !== 'false') st.push('current');
		if (el.required || aria('required') === 'true') st.push('required');
		if (el.readOnly || aria('readonly') === 'true') st.push('readonly');
		if (aria('invalid') === 'true') st.push('invalid');
		if (el.localName === 'dialog' && el.open) st.push('open');
		if (document.activeElement === el || (el.shadowRoot && el.shadowRoot.activeElement)) st.push('focused');
		return st;
	}

	function walk(node, parentStyle) {
		if (node.nodeType === Node.TEXT_NODE) {
			if (parentStyle && parentStyle.visibility !== 'visible') return [];
			return node.data ? [{ t: node.data }] : [];
		}
		if (node.nodeType !== Node.ELEMENT_NODE) return [];

		const el = node;
		const tag = el.localName;
		if (SKIP.has(tag)) return [];
		if (el.getAttribute('aria-hidden') === 'true' || el.hasAttribute('inert')) return [];
		if (tag === 'br') return [BR];

		const cs = getComputedStyle(el);
		if (cs.display === 'none') return [];

		if (tag === 'svg') {
			if (cs.visibility !== 'visible') return [];
			const t = el.querySelector(':scope > title');
			const n = collapse(el.getAttribute('aria-label') || (t && t.textContent));
			return n ? [{ r: 'img', n: clip(n, MAX_NAME) }] : [];
		}

		let role = (el.getAttribute('role') || '').trim().split(/\s+/)[0] || implicitRole(el);
		if (role === 'none' || role === 'presentation' || role === 'generic') role = null;
		if (!role && el.isContentEditable && !(el.parentElement && el.parentElement.isContentEditable)) role = 'textbox';

		const visibleSelf = cs.visibility === 'visible';
		let heuristic = false;
		if (!role && visibleSelf) {
			const tabbable = el.tabIndex >= 0 && el.hasAttribute('tabindex');
			const pointer = cs.cursor === 'pointer' && (!parentStyle || parentStyle.cursor !== 'pointer');
			if (tabbable || el.hasAttribute('onclick') || pointer) {
				role = 'clickable';
				heuristic = true;
			}
		}

		const isFormField = tag === 'input' || tag === 'textarea' || tag === 'select';
		const children = isFormField || tag === 'iframe' ? [] : walkChildren(el, cs);
		// A pointer cursor on a wrapper around real controls is not a control itself.
		if (heuristic && children.some(hasId)) role = null;

		if (!role || !visibleSelf) {
			const block = !cs.display.startsWith('inline') && cs.display !== 'contents';
			return block ? [BR, ...children, BR] : children;
		}

		const n = { r: role };
		const [name, fromContent] = nameOf(el, role);
		if (name) n.n = clip(name, MAX_NAME);
		if (INTERACTIVE.has(role) && !(tag === 'option' && el.closest('select'))) n.id = idFor(el);

		if (role === 'heading') {
			const lvl = /^h[1-6]$/.test(tag) ? +tag[1] : +el.getAttribute('aria-level');
			if (lvl) n.lvl = lvl;
		}
		if (role === 'link' && el.hasAttribute('href')) {
			const raw = el.getAttribute('href');
			if (raw && !raw.startsWith('javascript:') && raw !== '#') {
				let href = el.href;
				try {
					const u = new URL(el.href);
					if (u.origin === location.origin) href = u.pathname + u.search + u.hash;
				} catch (_) {}
				n.href = clip(href, 100);
			}
		}
		if (tag === 'iframe') n.href = clip(el.src || '', 100);

		if (tag === 'input' || tag === 'textarea') {
			const t = (el.getAttribute('type') || '').toLowerCase();
			if (!['button', 'submit', 'reset', 'image', 'checkbox', 'radio'].includes(t)) {
				if (el.value) n.v = t === 'password' ? '•'.repeat(Math.min(el.value.length, 12)) : clip(el.value, MAX_TEXT);
				const ph = collapse(el.getAttribute('placeholder'));
				if (ph && ph !== name) n.ph = clip(ph, MAX_NAME);
			}
		} else if (tag === 'select') {
			const opts = Array.from(el.options);
			const sel = opts.filter((o) => o.selected).map((o) => collapse(o.label || o.text));
			if (sel.length) n.v = clip(sel.join(', '), MAX_TEXT);
			n.c = opts.slice(0, MAX_OPTIONS).map((o) => {
				const on = { r: 'option', n: clip(collapse(o.label || o.text), MAX_NAME) };
				if (o.selected) on.st = ['selected'];
				if (o.disabled) on.st = (on.st || []).concat('disabled');
				return on;
			});
			if (opts.length > MAX_OPTIONS) n.c.push({ t: `… ${opts.length - MAX_OPTIONS} more options` });
		} else if (role === 'textbox' && el.isContentEditable) {
			const v = collapse(el.innerText);
			if (v) n.v = clip(v, MAX_TEXT);
		} else if (role === 'slider' || role === 'spinbutton' || role === 'progressbar' || role === 'meter') {
			const v = el.getAttribute('aria-valuetext') || el.getAttribute('aria-valuenow') || el.value;
			if (v !== undefined && v !== null && v !== '') n.v = String(v);
		}

		const st = states(el, role, cs);
		if (st.length) n.st = st.concat(n.st || []);

		const testid = el.getAttribute('data-testid') || el.getAttribute('data-test') || el.getAttribute('data-cy');
		if (testid) n.tid = clip(testid, 60);

		if (!n.c) {
			let c = normalize(children);
			// The name already contains the text content, so only keep what is still actionable.
			if (fromContent || (role === 'textbox' && el.isContentEditable)) {
				c = c.filter(hasId);
				// The kept children describe themselves, repeating their text in the name is noise.
				if (c.length && fromContent) delete n.n;
			}
			if (c.length) n.c = c;
		}
		return [n];
	}

	function hasId(n) {
		return !!n.r && (!!n.id || (n.c || []).some(hasId));
	}

	function normalize(items) {
		const out = [];
		let buf = '';
		const flush = () => {
			const t = collapse(buf);
			if (t) out.push({ t: clip(t, MAX_TEXT) });
			buf = '';
		};
		for (const it of items) {
			if (it.br) flush();
			else if (it.t !== undefined && !it.r) buf += it.t;
			else {
				flush();
				out.push(it);
			}
		}
		flush();
		return out;
	}

	const root = document.body || document.documentElement;
	const tree = root ? normalize(walkChildren(root, getComputedStyle(root))) : [];
	const se = document.scrollingElement || document.documentElement;
	return JSON.stringify({
		tree,
		next: S.next,
		url: location.href,
		title: document.title,
		sx: Math.round(window.scrollX),
		sy: Math.round(window.scrollY),
		sw: se ? se.scrollWidth : 0,
		sh: se ? se.scrollHeight : 0,
	});
}
